rs/nns/governance/src/voting.rs

```diff
@@ -0,0 +1,400 @@
+                    .neuron_would_follow_ballots(follower, self.topic, ballots)
+                {
+                    Ok(vote) => vote,
+                    Err(e) => {
+                        // This is a bad inconsistency, but there is
+                        // nothing that can be done about it at this
+                        // place.  We somehow have followers recorded that don't exist.
+                        eprintln!("error in cast_vote_and_cascade_follow when gathering induction votes: {:?}", e);
+                        Vote::Unspecified
+                    }
+                };
+                if vote != Vote::Unspecified {
+                    self.votes_to_cast.insert(follower, vote);
+                }
+            }
+        }
+
+        if !self.recent_neuron_ballots_to_record.is_empty() {
+            while let Some((neuron_id, vote)) = self.recent_neuron_ballots_to_record.pop_first() {
+                match neuron_store.with_neuron_mut(&neuron_id, |neuron| {
+                    neuron.register_recent_ballot(self.topic, &self.proposal_id, vote)
+                }) {
+                    Ok(_) => {
+                        self.votes_fully_cast.insert(neuron_id, vote);
+                    }
+                    Err(e) => {
+                        // This is a bad inconsistency, but there is
+                        // nothing that can be done about it at this
+                        // place.  We somehow have followers recorded that don't exist.
+                        eprintln!("error in cast_vote_and_cascade_follow when gathering induction votes: {:?}", e);
+                    }
+                };
+            }
+        }
+    }
+}
+
+// Retain only neurons that have a ballot that can still be cast.  This excludes
+// neurons with no ballots or ballots that have already been cast.
+fn retain_neurons_with_castable_ballots(
+    followers: &mut BTreeSet<NeuronId>,
+    ballots: &HashMap<u64, Ballot>,
+) {
+    followers.retain(|f| {
+        ballots
+            .get(&f.id)
+            // Only retain neurons with unspecified ballots
+            .map(|b| b.vote == Vote::Unspecified as i32)
+            // Neurons without ballots are also dropped
+            .unwrap_or_default()
+    });
+}
+
+pub(crate) fn cast_vote_and_cascade_follow(
+    proposal_id: &ProposalId,
+    ballots: &mut HashMap<u64, Ballot>,
+    voting_neuron_id: &NeuronId,
+    vote_of_neuron: Vote,
+    topic: Topic,
+    neuron_store: &mut NeuronStore,
+) {
+    let mut state_machine =
+        match VotingStateMachine::try_new(*proposal_id, topic, *voting_neuron_id, vote_of_neuron) {
+            Ok(sm) => sm,
+            Err(e) => {
+                // TODO - make this whole function return possible errors?
+                panic!("error in cast_vote_and_cascade_follow: {:?}", e);
+            }
+        };
+
+    while !state_machine.is_done() {
+        state_machine.continue_processing(neuron_store, ballots);
+    }
```

Comment:
I wasn't sure whether there is a guarantee that this while loop will eventually terminate. While we didn't have the guarantee before, a crucial difference is that the downside of an infinite loop will become much worse as soon as we introduce self-calls (previously, it will terminate by running out of instructions). IIUC if we introduce the self-calls and actually have an infinite loop, the governance canister cannot be stopped (correct me if I'm wrong).

After looking more closely, I am convinced that it will indeed terminate by applying the following reasoning:

Consider the following DAG where each node contains a set of neuron ids:
- unspecified ballot -> followers to check
- followers to check -> vote to cast
- vote to cast -> recent ballots
- vote to cast -> neurons to check followers

Each call to continue_processing will move at least one neuron along the DAG, or remove a neuron id from the DAG. Therefore since the neuron ids in the DAG is finite (mainly bounded by the ballots of the proposal), the algorithm must terminate. 

I tend to think this reasoning is useful otherwise in the future when we change the algorithm a little bit, it might be broken. Therefore I think it might be helpful to add some comments that reflects why the algorithm terminates.

Reply:
I added a test to make sure cyclic following couldn cause this to never terminate.  I wonder if that's a good enough test... what do you think?

Reply:
Indeed, the changes made it much easier to reason about the correctness and the test also helps. Thanks!